### ISSUE TITLE
The frame-level test ROC AUC micro score of TimeSformer model on UBnormal dataset is less than 68.5%

### ISSUE THREAD

**Issue by prdp2022** [lilygeorgescu/UBnormal#4]
Hi @lilygeorgescu,
Great Work

The TimeSformer model is fine-tuned on UBnormal dataset with frames sampled at 1/32, learning rate 0.0005 and 20 epochs. But the frame-level test ROC AUC micro is less than 68.5%. Can you please provide more information about the fine-tuning and testing of TimeSformer on UBnormal dataset.

**Comment by lilygeorgescu** [comment]
Hi,

Thank you for your interest in our work. I will tag my co-author @paulsum who was working on this experiment.

Thanks,
Best,
Lili

**Comment by paulsum** [comment]
Hi, 

As stated in the paper, the TimeSformer is trained on normal and abnormal actions. These actions are extracted from the original videos by dividing the videos in videos constructed from normal frames only and videos constructed from abnormal frames only (as per the frame-level ground truth information; an abnormal video is any video which contains at least one abnormal action, while a normal video is a video which contains no abnormal actions throughout its length) and feeding those videos to the model. 

So, normal videos are fed as they are to the model, while for abnormal videos we sometimes need to split them up into separate sub-videos. Take for example an abnormal video with 451 frames which contains one anomaly between frames 50 and 160. We would construct 3 videos as per the logic described above from the original one. The first normal video would range from frame 0 to frame 49, the second normal video would range from frame 161 to the end and finally an abnormal video would be constructed by looking only at frames 50 until frame 160.

After training was concluded, for inference we employed a rolling window approach. Here, each rolling window we feed to the model in order to obtain predictions will be exactly 256 frames (at least through experimentation this was found to be the optimal value). On each rolling window the model makes a prediction and outputs a score between 0 and 1. This score is assigned to all frames inside that rolling window. Then, with a step of 16 frames at a time, we take the next rolling window, iteratively passing the windows through the model. Naturally, there will be frames for which more than one score will be assigned in the end through this procedure. In order to obtain a single score per frame, we average all the predictions made for that frame in the video throughout inference. All these details describe what achieved the best results in our work after employing several other hyperparameter choices.

For videos shorter than 256 frames, we only have one rolling window, constructed as per the original logic in the TimeSformer repository. For longer videos, there also exists a special case when applying the rolling window mechanism. More specifically if we take the same 451 frame video and employ the rolling window approach, you will notice that towards the end, the so called "final" rolling window will stop at frame 448. In order to also cover the last 3 frames of the video, we take an additional window spanning from (end_frame-256 + 1) until end_frame. That being in this case from frame 196 to frame 451 (such that this window also has 256 frames in it).

**Comment by prdp2022** [comment]
Thank you @lilygeorgescu and @paulsum

**Comment by prdp2022** [comment]
Thank You @lilygeorgescu and @paulsum

**Comment by beautifulchoi** [comment]
@paulsum hello, thanks for your kindly explanation.
I wonder one another thing about this setting.
in the paper, default timesformer which can deal with 8 frame input is used.
So, at the inference phase, you mean make 256 frame one window and sliding each 8 frame and averaging them?(1-8, 2-9, 3-10 etc and averaging btw score), or not group each 8 frame and make prediction(1-8, 9-16, etc) or make 256 frame and sample which rate is 1/32? (256 frame to make 8 frame and deal it once)
